src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,152 @@
+        for fs in self.fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs['mdsmap'].get('standby_count_wanted')
+        return total
+
+    def get_current_standby_count(self, fs_name: str, daemons: Optional[List[orchestrator.DaemonDescription]] = None) -> int:
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        total = 0
+        if daemons is None:
+            return total
+        daemon_names = {
+            d.name() for d in daemons
+        }
+        for sb in self.fs_map['standbys']:
+            full_name = "mds.{}".format(sb['name'])
+            if full_name in daemon_names:
+                total += 1
+        return total
+
+    def get_current_active_count(self, fs_name: str):
+        assert self.fs_map is not None
+        for fsys in self.fs_map['filesystems']:
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return len(fsys.get('mdsmap').get('in'))
+        return 0
+
+    def verify_and_manage_mds_instance(self, fs_name: str):
+        assert self.fs_map is not None
+
+        daemons = self.get_daemons(fs_name)
+        standbys_required = self.get_required_standby_count(fs_name)
+        standbys_current = self.get_current_standby_count(fs_name, daemons)
+        active = self.get_current_active_count(fs_name)
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}"
+                      .format(standbys_required, standbys_current))
+        if standbys_required is None:
+            return
+        total = standbys_current + active
+        if standbys_current > standbys_required:
+            # remove one mds at a time and wait for updated fs_map
+            self.log.info("killing {} standby mds ..."
+                          .format((standbys_current - standbys_required)))
+            total -= standbys_current - standbys_required
+        elif standbys_current < standbys_required:
+            # launch one mds at a time and wait for updated fs_map
+            self.log.info("spawning {} standby mds ..."
+                          .format((standbys_required - standbys_current)))
+            total += standbys_required - standbys_current
+
+        newspec = self.update_daemon_count(fs_name, total)
+        # newspec == None means that the placement count is already at the
+        # required count
+        if newspec is None:
+            return
+
+        self.log.info("new placement count:{0}".format(newspec.placement.count))
+
+        completion = self.apply_mds(newspec)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def notify(self, notify_type, notify_id):
+        # we don't know for which fs config has been changed
+        if notify_type != 'fs_map':
+            return
+        self.fs_map = self.get('fs_map')
```

Comment:
I'm not seeing a reason to persist the fs_map in `self.fs_map`. Just pass it as a paramter?

Reply:
I'll change this to pass fs_map as a parameter